src/v/cloud_storage/remote_partition.cc

```diff
@@ -905,10 +905,18 @@ ss::future<remote_partition::erase_result> remote_partition::erase(
     // and construct a special one.
     retry_chain_node local_rtc(as, erase_timeout, erase_backoff);
 
+    auto replaced_segments = manifest.replaced_segments();
+
     // Erase all segments
+    // Loop over normal segments and replaced (pending deletion) segments
+    // at the same time, to batch them together into as few DeleteObjects
+    // requests as possible
     const size_t batch_size = 1000;
     auto segment_i = manifest.begin();
-    while (segment_i != manifest.end()) {
+    auto replaced_i = replaced_segments.begin();
+
+    while (segment_i != manifest.end()
+           || replaced_i != replaced_segments.end()) {
```

Comment:
nit: some sort of range iteration would be nice here, but I don't think this works out the box
```
for (const auto &seg : {manifest, replaced_segments} | std::views::join) { ...}
```